### PR TITLE
[ML] Logistic regression loss function for boosted tree training

### PR DESCRIPTION
**Pull request by tveasey**
This implements binomial logistic regression for the boosted tree. In particular, this targets cross entropy and builds a forest to predict the class log-odds.

We should also have been including sum square leaf weight penalty in the calculation of the optimum tree leaf values, since the splits are chosen targeting the regularised objective. (Note that the regularisation applies to the log-odds for logistic regression, i.e. we'll shrink the log-odds towards zero and so the predicted probabilities towards 0.5.)

I haven't wired this in yet, since that work depends on #701.

### CODE REVIEW

File: lib/maths/unittest/CBoostedTreeTest.cc [hunk 1]
```diff
@@ -802,13 +818,225 @@ void CBoostedTreeTest::testDepthBasedRegularization() {
...
+            TDoubleVec noise;
+            rng.generateNormalSamples(0.0, 1.0, rows, noise);
+            return [=](const TRowRef& row) {
+                double x{0.0};
+                for (std::size_t i = 0; i < cols - 1; ++i) {
+                    x += weights[i] * row[i];
+                }
+                return maths::CTools::logisticFunction(x + noise[row.index()]);
+            };
+        }();
```
Comment (valeriy42) [inline lib/maths/unittest/CBoostedTreeTest.cc]: There isn't actually need in defining this as an anonymous lambda function, is there? Nesting of multiple lambda functions here hinders readability. Same also below.

File: include/maths/CBoostedTree.h [hunk 2]
```diff
@@ -64,6 +118,9 @@ class MATHS_EXPORT CArgMinLoss {
     CArgMinLoss& operator=(const CArgMinLoss& other);
     CArgMinLoss& operator=(CArgMinLoss&& other) = default;
 
+    //! The number of passes over the data this needs.
+    bool nextPass() const;
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTree.h]: +1. This comment is just wrong and reflects an old version, I'll update.

File: include/maths/CBoostedTree.h [hunk 3]
```diff
@@ -51,6 +65,46 @@ class MATHS_EXPORT CArgMinMseImpl final : public CArgMinLossImpl {
 private:
     TMeanAccumulator m_MeanError;
 };
+
+//! \brief Finds the value to add to a set of predicted log-odds which minimises
+//! regularised the cross entropy loss w.r.t. the actual categories.
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTree.h]: See [c1b3ec1](https://github.com/elastic/ml-cpp/pull/713/commits/c1b3ec184f1cd7f590b33bfe3cdf9af2da4b00b0)

File: lib/maths/CBoostedTree.cc [hunk 4]
```diff
@@ -43,7 +62,118 @@ void CArgMinMseImpl::merge(const CArgMinLossImpl& other) {
 }
 
 double CArgMinMseImpl::value() const {
-    return CBasicStatistics::mean(m_MeanError);
+    double count{CBasicStatistics::count(m_MeanError)};
+    return count == 0.0
+               ? 0.0
+               : count / (count + this->lambda()) * CBasicStatistics::mean(m_MeanError);
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: I added a detailed [explanation](https://github.com/elastic/ml-cpp/pull/713/commits/1b9c6a8fc1cfc34da0bf2a464b1b00de73655ee3)

File: include/maths/CBoostedTree.h [hunk 5]
```diff
@@ -51,6 +65,46 @@ class MATHS_EXPORT CArgMinMseImpl final : public CArgMinLossImpl {
...
+    }
+
+    double bucketWidth() const {
+        return m_MinMaxPrediction.range() /
+               static_cast<double>(m_BucketCategoryCounts.size());
+    }
+
+private:
+    std::size_t m_CurrentPass = 0;
+    TMinMaxAccumulator m_MinMaxPrediction;
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTree.h]: Better :). See [f011036](https://github.com/elastic/ml-cpp/pull/713/commits/f01103699aba8c7e5a2ac7115863156d210602cc).

File: include/maths/CBoostedTree.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTree.h]: See [dd77bb3](https://github.com/elastic/ml-cpp/pull/713/commits/dd77bb3018a63a04280113f8e88a3e98535aefba).

File: lib/maths/CBoostedTree.cc [hunk 6]
```diff
@@ -43,7 +62,118 @@ void CArgMinMseImpl::merge(const CArgMinLossImpl& other) {
...
+    }
+}
+
+double CArgMinLogisticImpl::value() const {
+
+    std::function<double(double)> objective;
+    double minWeight;
+    double maxWeight;
+
+    if (this->bucketWidth() == 0.0) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: See [2b88359](https://github.com/elastic/ml-cpp/pull/713/commits/2b8835985a65ac7ae1bd5512a56c2d2b00cca1cc).

File: lib/maths/unittest/CBoostedTreeTest.cc [hunk 7]
```diff
@@ -802,13 +818,225 @@ void CBoostedTreeTest::testDepthBasedRegularization() {
...
+                rng.generateNormalSamples(label, 2.0, 1, weight);
+                weights.push_back(weight[0]);
+                min = std::min(min, weight[0]);
+                max = std::max(max, weight[0]);
+            }
+
+            double expected;
+            double objectiveAtExpected;
+            std::size_t maxIterations{20};
+            maths::CSolvers::minimize(-max, -min, objective(-max), objective(-min),
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/unittest/CBoostedTreeTest.cc]: I added a comment (and tweaked range we use slightly). See [4cb6f0d](https://github.com/elastic/ml-cpp/pull/713/commits/4cb6f0df3e3627907e14f833a5c143df62fe1e90).

File: lib/maths/unittest/CBoostedTreeTest.cc [hunk 8]
```diff
@@ -802,13 +818,225 @@ void CBoostedTreeTest::testDepthBasedRegularization() {
...
+            CPPUNIT_ASSERT_DOUBLES_EQUAL(objectiveAtExpected, objective(actual),
+                                         1e-5 * objectiveAtExpected);
+        }
+    }
+}
+
+void CBoostedTreeTest::testLogisticRegression() {
+
+    // Test we approximately minimise the cross entropy if the category labels
+    // are generated from log odds which are a linear function of the regressors.
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/unittest/CBoostedTreeTest.cc]: I [rewrote](https://github.com/elastic/ml-cpp/pull/713/commits/7c3dbcfa3bea3537778c64c8bbfe0cc9bc2de060). Hopefully, clearer.

File: lib/maths/unittest/CBoostedTreeTest.cc [hunk 9]
```diff
@@ -802,13 +818,225 @@ void CBoostedTreeTest::testDepthBasedRegularization() {
...
+            double loss{0.0};
+            for (std::size_t i = 0; i < labels.size(); ++i) {
+                double p{maths::CTools::logisticFunction(weights[i] + weight)};
+                loss -= (1.0 - labels[i]) * maths::CTools::fastLog(1.0 - p) +
+                        labels[i] * maths::CTools::fastLog(p);
+            }
+            return loss + lambda * maths::CTools::pow2(weight);
+        };
+
+        for (std::size_t t = 0; t < 10; ++t) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/unittest/CBoostedTreeTest.cc]: See [4b62cac](https://github.com/elastic/ml-cpp/pull/713/commits/4b62cac0d0c2963c1c0b1b5a6f8fccd2b23f79eb).

File: lib/maths/CBoostedTree.cc [hunk 10]
```diff
@@ -96,15 +234,50 @@ bool CMse::isCurvatureConstant() const {
...
 const std::string CMse::NAME{"mse"};
+
+std::unique_ptr<CLoss> CLogistic::clone() const {
+    return std::make_unique<CLogistic>(*this);
+}
+
+double CLogistic::value(double prediction, double actual) const {
+    // Cross entropy
+    prediction = CTools::logisticFunction(prediction);
+    return -((1.0 - actual) * CTools::fastLog(1.0 - prediction) +
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: As discussed offline, this is actually managed elsewhere, i.e. lambda and gamma are explicitly used in the code which evaluates candidate splits of the tree. This loss function is thus the unregularised version.

File: lib/maths/CBoostedTree.cc [hunk 11]
```diff
@@ -43,7 +62,118 @@ void CArgMinMseImpl::merge(const CArgMinLossImpl& other) {
...
+                loss -= static_cast<double>(c0) * CTools::fastLog(1.0 - p) +
+                        static_cast<double>(c1) * CTools::fastLog(p);
+            }
+            return loss + this->lambda() * CTools::pow2(weight);
+        };
+
+        // Choose a weight interval in which all probabilites vary from close to
+        // zero to close to one.
+        minWeight = -m_MinMaxPrediction.max() - 2.0;
+        maxWeight = -m_MinMaxPrediction.min() + 2.0;
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: I added a longer comment in [this](https://github.com/elastic/ml-cpp/pull/713/commits/4cb6f0df3e3627907e14f833a5c143df62fe1e90) commit.

File: lib/maths/CBoostedTreeImpl.cc [hunk 12]
```diff
@@ -433,7 +433,6 @@ CBoostedTreeImpl::crossValidateForest(core::CDataFrame& frame,
         lossMoments.add(loss);
         LOG_TRACE(<< "fold = " << i << " forest size = " << forest.size()
                   << " test set loss = " << loss);
-        m_TrainingProgress.increment();
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: This got moved into trainForest call just above. This is called directly elsewhere where it wasn't monitored.

File: lib/maths/CBoostedTree.cc [hunk 13]
```diff
@@ -43,7 +62,118 @@ void CArgMinMseImpl::merge(const CArgMinLossImpl& other) {
...
+            return this->lambda() * CTools::pow2(weight) -
+                   static_cast<double>(c0) * CTools::fastLog(1.0 - p) -
+                   static_cast<double>(c1) * CTools::fastLog(p);
+        };
+
+        // Weight shrinkage means the optimal weight will be somewhere
+        // between the logit of the empirical probability and zero.
+        std::size_t c0{m_CategoryCounts(0) + 1};
+        std::size_t c1{m_CategoryCounts(1) + 1};
+        double p{static_cast<double>(c1) / static_cast<double>(c0 + c1)};
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: See [this](https://github.com/elastic/ml-cpp/pull/713/commits/85aa19234abeebe7cf476b7321c5feafc7efad5c) commit.

File: lib/maths/CBoostedTree.cc [hunk 14]
```diff
@@ -62,6 +62,15 @@ void CArgMinMseImpl::merge(const CArgMinLossImpl& other) {
...
 double CArgMinMseImpl::value() const {
+
+    // We searching for the value x which minimises
+    //
+    //    x^* = argmin_x{ sum_i{(a_i - (p_i + x))^2} + lambda * x^2 }
+    //
+    // This is convex so there is one minimum where derivative w.r.t. x is zero
+    // and x^* = 1 / (n + lambda) sum_i{ a_i - p_i }. Denoting the mean prediction
+    // error m = 1/n sum_i{ a_i - p_i } we have x^* = n / (n + lambda) m.
+
```
Comment (valeriy42) [inline lib/maths/CBoostedTree.cc]: Good job on explaining what the function does! :+1:

File: lib/maths/CBoostedTree.cc [hunk 15]
```diff
@@ -124,6 +133,9 @@ double CArgMinLogisticImpl::value() const {
     double minWeight;
     double maxWeight;
 
+    // This is true if and only if all the predictions were identical. In this
+    // case we only need one pass over the data and can compute the optimal
```
Comment (valeriy42) [inline lib/maths/CBoostedTree.cc]: :+1:

File: lib/maths/CBoostedTree.cc [hunk 16]
```diff
@@ -157,9 +172,12 @@ double CArgMinLogisticImpl::value() const {
         };
 
         // Choose a weight interval in which all probabilites vary from close to
-        // zero to close to one.
-        minWeight = -m_MinMaxPrediction.max() - 2.0;
-        maxWeight = -m_MinMaxPrediction.min() + 2.0;
+        // zero to close to one. In particular, the idea is to minimize the leaf
+        // weight on an interval [a, b] where if we add "a" the log-odds for all
+        // rows <= -5, i.e. max prediction + a = -5, and if we add "b" the log-odds
+        // for all rows >= 5, i.e. min prediction + a = 5.
```
Comment (valeriy42) [inline lib/maths/CBoostedTree.cc]: Nice explanation! :+1:

File: lib/maths/unittest/CBoostedTreeTest.cc [hunk 17]
```diff
@@ -908,6 +910,11 @@ void CBoostedTreeTest::testLogisticMinimizer() {
                 max = std::max(max, weight[0]);
             }
 
+            // Choose a weight interval in which all probabilites vary from close to
+            // zero to close which we know will contain the true optimum. The idea is
+            // to minimize the leaf weight on an interval [a, b] where if we add "a"
+            // the log-odds for all rows <= 0, i.e. max prediction + a = 0, and if we
+            // add "b" the log-odds for all rows >= 0, i.e. min prediction + a = 0.
```
Comment (valeriy42) [inline lib/maths/unittest/CBoostedTreeTest.cc]: This comment seems to be out of place here.

File: lib/maths/unittest/CBoostedTreeTest.cc (hunk 1 above)
Comment (tveasey) [reply lib/maths/unittest/CBoostedTreeTest.cc]: Done

**Comment by tveasey** [comment]
I've now addressed all your review comments. Can you take another look @valeriy42.

File: lib/maths/unittest/CBoostedTreeTest.cc [hunk 18]
```diff
@@ -956,8 +963,17 @@ void CBoostedTreeTest::testLogisticMinimizer() {
...
 
-    // Test we approximately minimise the cross entropy if the category labels
-    // are generated from log odds which are a linear function of the regressors.
+    // The idea of this test is to create a random linear relationship between
+    // the feature values and the log-odds of each class, i.e.
+    //
+    //   log-odds(class_1) = sum_i{ w * x_i }
+    //
+    // where, w is some fixed weight vector and x_i denoted the i'th feature vector.
+    // We are try to recover this relationship in logistic regression by observing
```
Comment (valeriy42) [inline lib/maths/unittest/CBoostedTreeTest.cc]:
```suggestion
    // We try to recover this relationship in logistic regression by observing
```

File: lib/maths/unittest/CBoostedTreeTest.cc (hunk 17 above)
Comment (tveasey) [reply lib/maths/unittest/CBoostedTreeTest.cc]: It was just that you'd commented on searching [-max, -min] here as well, but perhaps no need to repeat. I'll just remove.
